Data store interface for v3 API reports endpoints.

peterfabian opened the pull request:
I tried to follow a similar structure as we have for other data stores, Claudiu mentioned `WC_Product_Query` as an example, so I took the inspiration from that one.
I created a class 
`WC_Reports_Data_Store` as a common ancestor and one child class `WC_Reports_Revenue_Data_Store` as an example. 
Also defined a super simple common interface `WC_Reports_Data_Store_Interface `.

The idea was that the query class  `WC_Reports_Revenue_Query` then interfaces with the data store class and simply requests the data specified by the parameters from the user. It should also describe the default parameters accepted from the user/other code.

If it's too much abstraction, please let me know and I can try to reorganize things a bit.

I wasn't 100% sure how the order_stats table will work, so I assumed there might be some missing time intervals (I guess it could be quite wasteful to store all 0 values in case there were no orders, especially for smaller shops).
Therefore, we might need a solution to 'fill in' the missing time intervals with 0 data. As I see it, there are 2 solutions--either to fill the gaps in SQL query or fill the gaps in PHP. I opted for the latter, as I think PHP is a bit more flexible and if we used the former approach, we would need to create a sequencing virtual table each time we would run a query. This table would have to be then joined with actual data and can only have fixed size, which I don't like (solution is more thoroughly described in [this blog post](https://www.plumislandmedia.net/mysql/filling-missing-data-sequences-cardinal-integers/)).
The drawback of the PHP based solution is that it requires some more processing--one extra db query to count the number of rows in the db within the time period and using some arithmetic to correctly calculate offset in the LIMIT clause, plus sorting and filtering after the SQL query (as new 0 records are added to the array of intervals). 

If you feel we should switch to SQL solution, please let me know in comments. 

If this assumption is not true and we will have data for every hour of every day, then we can make things easier and make the code shorter and easier to understand.

Nevertheless, week numbering is a bit tricky. I went with two versions:
- one that is ISO 8601 compliant, that is supported by both PHP and MySQL -- if the week start is set to Monday
- the other where the first week of the year always starts on January 1st, then next week starts on the following [day of beginning of the week].

Pending: 
- [ ] tests

peterfabian committed:
Initial commit of data store interface for reports endpoints.

peterfabian committed:
Moved the example usage to correct class.

claudiosanches commented on includes/class-wc-reports-interval.php:
```diff
@@ -0,0 +1,227 @@
+defined( 'ABSPATH' ) || exit;
+
+define( 'SECONDS_PER_DAY', 86400 );
+define( 'SECONDS_PER_HOUR', 3600 );
```
WordPress already declare constants for it: https://github.com/WordPress/WordPress/blob/1ae9491548e64e66c3283c97dd082a691fa57e08/wp-includes/default-constants.php#L116-L134

peterfabian replied:
fixed

peterfabian committed:
Use standard WP time constants instead of new ones.

peterfabian committed:
Next month and quarter made year-end aware.

peterfabian committed:
Harmonized week calculation between PHP and MySQL.

- Updated week calcs to ISO in case of Monday week start and simple week numbering otherwise.
- Fixed handling of data when all records are present in the db.

claudiulodro commented on includes/class-wc-reports-interval.php:
```diff
@@ -0,0 +1,254 @@
+	 * @return int|null
+	 */
+	public static function quarter( $datetime ) {
+		// TODO: is there a smarter way? Is floor((m - 1)/3) + 1 better?
```
I like the current way. It's easier to read. (better performance also :P)

claudiulodro reviewed (approved):
This looks good to me. I like that it follows a consistent design pattern with the rest of the codebase. 👍 

We can fine-tune the internals further once we've got the endpoints set up. Let's get this merged so we aren't blocking next week's sprint items.

@peterfabian how were you thinking this should integrate with [the lookup table](https://github.com/woocommerce/woocommerce/pull/20581/files#diff-70b4939f537887a27843503488e23d44R20)? Looking at the data store I think it probably makes sense to get rid of `WC_Order_Stats` and make all that part of `WC_Reports_Revenue_Data_Store`? (After this is merged)
